File: lib/rgame/util/solid_grid.rb
# frozen_string_literal: true

require 'rgame/util_ext'

module RGame
  module Util
    # Which cells of a tile grid are solid: one byte a cell, in C. A value with no handle
    # behind it, which is why it lives in Util and why the engine layer may own one.
    #
    # It is the one store a tile world's solidity lives in. The blockers that stop a walker
    # and the RouteSearch that plans one both read it, so they cannot disagree about a wall,
    # and a cell changed with #set_solid is changed for both at once.
    #
    # #revision counts the changes: it moves on a #set_solid that changes a cell and on no
    # other, so anything derived from the cells — a search's region labels — can tell
    # whether it is stale without being told.
    #
    # Coordinates are Integers, and anything else is a TypeError. Outside the grid is open,
    # including an Integer too large to be inside any grid.
    #
    #   grid = RGame::Util::SolidGrid.build(8, 3) { |col, _row| col == 3 }
    #   grid.solid?(3, 1)    # => true
    #   grid.solid?(-1, 0)   # => false
    #   grid.set_solid(3, 1, false)
    #   grid.revision        # => 1
    class SolidGrid
      # A grid with each cell's solidity read from the block, once per cell, row by row.
      def self.build(width, height)
        grid = new(width, height)
        height.times do |row|
          width.times { |col| grid.set_solid(col, row, true) if yield(col, row) }
        end
        grid
      end
    end
  end
end

File: ext/rgame_util/solid_grid.c
/*
 * solid_grid.c — the solidity store of a tile grid. Pure; see solid_grid.h.
 */

#include "solid_grid.h"

#include <stdlib.h>

bool rgame_solid_grid_size_ok(int64_t width, int64_t height) {
    return width >= 0 && height >= 0 && (height == 0 || width <= INT32_MAX / height);
}

bool rgame_solid_grid_init(rgame_solid_grid *grid, int32_t width, int32_t height) {
    if (!rgame_solid_grid_size_ok(width, height)) {
        return false;
    }
    size_t count = (size_t)width * (size_t)height;
    uint8_t *cells = NULL;
    if (count > 0) {
        cells = calloc(count, 1);
        if (!cells) {
            return false;
        }
    }
    grid->width = width;
    grid->height = height;
    grid->cells = cells;
    grid->revision = 0;
    return true;
}

void rgame_solid_grid_free(rgame_solid_grid *grid) {
    free(grid->cells);
    grid->cells = NULL;
    grid->width = 0;
    grid->height = 0;
}

bool rgame_solid_grid_inside(const rgame_solid_grid *grid, int32_t col, int32_t row) {
    return col >= 0 && row >= 0 && col < grid->width && row < grid->height;
}

bool rgame_solid_grid_solid(const rgame_solid_grid *grid, int32_t col, int32_t row) {
    return rgame_solid_grid_inside(grid, col, row) && grid->cells[row * grid->width + col];
}

bool rgame_solid_grid_set(rgame_solid_grid *grid, int32_t col, int32_t row, bool solid) {
    if (!rgame_solid_grid_inside(grid, col, row)) {
        return false;
    }
    uint8_t *cell = &grid->cells[row * grid->width + col];
    uint8_t value = solid ? 1 : 0;
    if (*cell != value) {
        *cell = value;
        grid->revision++;
    }
    return true;
}
